File: python/src/ai_sdk/api/personas.py
"""Personas namespace."""

from __future__ import annotations

import builtins
from collections.abc import Callable
from typing import Any
from urllib.parse import quote

from ai_sdk._http import AsyncHTTPClient, HTTPClient
from ai_sdk.exceptions import AISdkError, PersonaNotFoundError
from ai_sdk.models import CreatePersonaRequest, PersonaInfo

_PersonaInfoList = builtins.list[PersonaInfo]


class PersonasAPI:
    """Namespace for persona operations."""

    def __init__(self, http: HTTPClient, async_http: AsyncHTTPClient | None) -> None:
        self._http = http
        self._async_http = async_http

    def list(self, limit: int | None = None) -> _PersonaInfoList:
        return _paginate(self._http, "/", PersonaInfo.from_dict, limit=limit)

    async def alist(self, limit: int | None = None) -> _PersonaInfoList:
        if self._async_http is None:
            raise RuntimeError(
                "Async HTTP client not available. "
                "Use AISdk with enable_async=True for async operations."
            )
        return await _apaginate(self._async_http, "/", PersonaInfo.from_dict, limit=limit)

    def get(self, name: str) -> PersonaInfo:
        try:
            encoded_name = quote(name, safe="")
            response = self._http.get(f"/name/{encoded_name}")
            return PersonaInfo.from_dict(response)
        except AISdkError as e:
            if e.status_code == 404:
                raise PersonaNotFoundError(name) from e
            raise

    async def aget(self, name: str) -> PersonaInfo:
        if self._async_http is None:
            raise RuntimeError(
                "Async HTTP client not available. "
                "Use AISdk with enable_async=True for async operations."
            )
        try:
            encoded_name = quote(name, safe="")
            response = await self._async_http.get(f"/name/{encoded_name}")
            return PersonaInfo.from_dict(response)
        except AISdkError as e:
            if e.status_code == 404:
                raise PersonaNotFoundError(name) from e
            raise

    def create(self, request: CreatePersonaRequest) -> PersonaInfo:
        response = self._http.post("/", json=request.to_api_dict())
        return PersonaInfo.from_dict(response)

    async def acreate(self, request: CreatePersonaRequest) -> PersonaInfo:
        if self._async_http is None:
            raise RuntimeError(
                "Async HTTP client not available. "
                "Use AISdk with enable_async=True for async operations."
            )
        response = await self._async_http.post("/", json=request.to_api_dict())
        return PersonaInfo.from_dict(response)


def _paginate(
    http: HTTPClient,
    path: str,
    mapper: Callable[[dict[str, Any]], Any],
    limit: int | None = None,
    page_size: int = 100,
) -> builtins.list:
    results: builtins.list = []
    after: str | None = None
    while True:
        params: dict[str, Any] = {"limit": page_size}
        if after:
            params["after"] = after
        response = http.get(path, params=params)
        for item in response.get("data", []):
            results.append(mapper(item))
            if limit is not None and len(results) >= limit:
                return results[:limit]
        paging = response.get("paging") or {}
        after = paging.get("after")
        if not after:
            break
    return results


async def _apaginate(
    http: AsyncHTTPClient,
    path: str,
    mapper: Callable[[dict[str, Any]], Any],
    limit: int | None = None,
    page_size: int = 100,
) -> builtins.list:
    results: builtins.list = []
    after: str | None = None
    while True:
        params: dict[str, Any] = {"limit": page_size}
        if after:
            params["after"] = after
        response = await http.get(path, params=params)
        for item in response.get("data", []):
            results.append(mapper(item))
            if limit is not None and len(results) >= limit:
                return results[:limit]
        paging = response.get("paging") or {}
        after = paging.get("after")
        if not after:
            break
    return results

File: typescript/src/api/personas.ts
/**
 * Personas namespace.
 *
 * Backed by an HttpClient rooted at /api/v1/agents/personas.
 */

import type { HttpClient } from '../http.js';
import type { PaginatedResponse } from '../models.js';
import type { CreatePersonaRequest, PersonaInfo } from '../types.js';

/**
 * Options for list operations on the personas namespace.
 */
export interface PersonasListOptions {
  /** Maximum number of personas to return. If omitted, returns all personas. */
  limit?: number;
}

/**
 * API response for a persona.
 * @internal
 */
interface ApiPersonaInfo {
  id: string;
  name: string;
  displayName?: string;
  description?: string;
  prompt?: string;
  provider: string;
}

function mapPersonaInfo(data: ApiPersonaInfo): PersonaInfo {
  return {
    id: data.id,
    name: data.name,
    displayName: data.displayName,
    description: data.description,
    prompt: data.prompt,
    provider: data.provider,
  };
}

/**
 * Namespace for persona operations.
 */
export class PersonasApi {
  private readonly http: HttpClient;

  constructor(http: HttpClient) {
    this.http = http;
  }

  /**
   * List all personas.
   *
   * Automatically paginates through all results.
   */
  async list(options?: PersonasListOptions): Promise<PersonaInfo[]> {
    return paginate<ApiPersonaInfo, PersonaInfo>(
      (params) => this.http.get<PaginatedResponse<ApiPersonaInfo>>('/', params),
      mapPersonaInfo,
      options?.limit
    );
  }

  /**
   * Get a persona by name.
   *
   * @throws {PersonaNotFoundError} If the persona is not found.
   */
  async get(name: string): Promise<PersonaInfo> {
    const response = await this.http.get<ApiPersonaInfo>(
      `/name/${encodeURIComponent(name)}`,
      undefined,
      undefined,
      'persona',
      name
    );
    return mapPersonaInfo(response);
  }

  /**
   * Create a new persona.
   */
  async create(request: CreatePersonaRequest): Promise<PersonaInfo> {
    const body = {
      name: request.name,
      description: request.description,
      prompt: request.prompt,
      displayName: request.displayName,
      provider: request.provider ?? 'user',
      owners: request.owners,
    };

    const response = await this.http.post<ApiPersonaInfo>('/', body);
    return mapPersonaInfo(response);
  }
}

/**
 * Paginate through all results from a list endpoint.
 * @internal
 */
async function paginate<T, U>(
  fetcher: (
    params: Record<string, string | number | boolean>
  ) => Promise<PaginatedResponse<T>>,
  mapper: (item: T) => U,
  limit?: number,
  pageSize: number = 100
): Promise<U[]> {
  const results: U[] = [];
  let after: string | undefined = undefined;

  // eslint-disable-next-line no-constant-condition
  while (true) {
    const params: Record<string, string | number | boolean> = { limit: pageSize };
    if (after) {
      params.after = after;
    }

    const response = await fetcher(params);
    const data = response.data || [];
    results.push(...data.map(mapper));

    if (limit !== undefined && results.length >= limit) {
      return results.slice(0, limit);
    }

    after = response.paging?.after;
    if (!after) {
      break;
    }
  }

  return results;
}
